File: services/comfortService/comfortService.cpp
#include <string>

#include <grpcpp/grpcpp.h>
#include <grpc/grpc.h>
#include <grpcpp/server.h>
#include <grpcpp/server_builder.h>
#include <grpcpp/server_context.h>
#include "proto/v1/generated/comfort_service_api_v1.pb.h"
#include "proto/v1/generated/comfort_service_api_v1.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;

using comfortServiceAPI::v1::ComfortService;
using comfortServiceAPI::v1::ComfortRequest;
using comfortServiceAPI::v1::ComfortResponse;
using comfortServiceAPI::v1::VDVRequest;
using comfortServiceAPI::v1::VDVResponse;

class ComfortServiceImplementation final : public ComfortService::Service {
    /* 'Comfort Service' offers three service calls that provide information about human comfort onboard, in response to vessel vibrations.
    */

    enum evaluationCase {Passenger, Crew, Work};

    void EvaluateHabitability (float acceleration, evaluationCase area, ComfortResponse* response)
    {
        /* This function takes an acceleration, an analysis area, and the response message it will be modifying as inputs. It evaluates the habitibility of a vessel using the acceleration input, based on what zone it falls within according to ISO 6954:2000.
        */

        int upperLimit;
        int lowerLimit;

        // Set the upper and lower limits for evaluation, based on the selected evaluation area.
        switch (area)
        {
            case Passenger:
                upperLimit = 4;
                lowerLimit = 2;
                break;
            case Crew:
                upperLimit = 6;
                lowerLimit = 3;
                break;
            case Work:
                upperLimit = 8;
                lowerLimit = 4;
                break;
            default:
                // Default to the most sensitive case (passenger cabins)
                upperLimit = 4;
                lowerLimit = 2;
                break;
        }
    
        // Evaluate habitibility
        if (acceleration < (float) lowerLimit)
        {
            // If acceleration is less than the lower limit, adverse commetns are unlikely.
            response->set_rating(comfortServiceAPI::v1::ADVERSE_COMMENTS_UNLIKELY);
        } else if (acceleration < upperLimit)
        {
            // Acceleration in the zone between the lower and upper limits is considered to be commonly accepted operating conditions.
            response->set_rating(comfortServiceAPI::v1::NORMAL_OPERATING_CONDITIONS);

        } else
        {
            // If acceleration is greater than the upper limit, adverse comments are probable.
            response->set_rating(comfortServiceAPI::v1::ADVERSE_COMMENTS_PROBABLE);
        }

    }
    
    void AssessComfort (double equivalentAcceleration, ComfortResponse* response)
    {
        if(equivalentAcceleration < 0.315)
        {
            response->set_rating(comfortServiceAPI::v1::NOT_UNCOMFORTABLE);
        } else if (equivalentAcceleration < 0.6)
        {
            response->set_rating(comfortServiceAPI::v1::SLIGHTLY_UNCOMFORTABLE);
        } else if (equivalentAcceleration < 0.9)
        {
            response->set_rating(comfortServiceAPI::v1::FAIRLY_UNCOMFORTABLE);
        } else if (equivalentAcceleration < 1.4)
        {
            response->set_rating(comfortServiceAPI::v1::UNCOMFORTABLE);
        } else if (equivalentAcceleration < 2)
        {
            response->set_rating(comfortServiceAPI::v1::VERY_UNCOMFORTABLE);
        } else
        {
            response->set_rating(comfortServiceAPI::v1::EXTREMELY_UNCOMFORTABLE);
        }
    }

    double calculateEquivalentVibration(double timeStamps[], double weightedVibration[])
    {
        /* This function takes a time series of weighted vibration data as well as their assosciated timestamps. Using these, it calculates the equivalent vibration magnitude according to SANS 2631-1 (Appendix C, equ C.1).
        */

        float timeGap = 0; // This variable holds the time difference between vibration samples
        double numerator = 0; // This variables holds the accumulated numerator use in the calculation of the equivalent vibration magnitude
        double denominator = 0; // This variables holds the accumulated denominator use in the calculation of the equivalent vibration magnitude
        
        // Iterate through the request samples
        for(int i = 1; i < sizeof(weightedVibration); i++)
        {
            // Find the time difference between estimates. Assume the acceleration signal is constant over this time period
            timeGap = timeStamps[i] - timeStamps[i-1];

            // Increment numerator of the equivalent vibration equation
            numerator += ( std::pow(weightedVibration[i],2)*timeGap );

            // Increment denominator of the equivalent vibration equation
            denominator += timeGap;
        }

        // Calculate and return equivalent vibration
        return sqrt(numerator/denominator);
    }

    Status ComfortRating(
        ServerContext* context, 
        const ComfortRequest* request, 
        ComfortResponse* response
    ) override {
        /* The 'Comfort Rating' call provides foresight for ?? decision-making by providing a comfort rating for a proposed route, based on estimated vibrations on board. 
        */

        double timeStampArray[request->unix_time_size()];
        double vibrationXArray[request->human_weighted_vibration_x_size()];

        // Unpack and add all timestamps to the response message
        for(int i = 0; i < request->unix_time_size(); i++)
        {
            timeStampArray[i] = request->unix_time(i); // Unpack the request message into a local array
            response->set_unix_time(i, timeStampArray[i]); // Set response message time
        }

        // Unpack all human weighted vibrations
        for(int i = 0; i < request->human_weighted_vibration_x(); i++_)
        {
            vibrationXArray[i] = request->human_weighted_vibration_x(i);
        }

        // ________Calculate equivalent vibration________

        double equivalentVibration = calculateEquivalentVibration(timeStampArray, vibrationXArray);

        // 2. Calculate VDV
        double vdvAccelerationZ = 1.4 * equivalentVibration * std::pow((timeStampArray[sizeof(timeStampArray) - 1] - timeStampArray[0]), 0.25);

        // 3. Assess comfort
        AssessComfort(equivalentVibration, response);

        return Status::OK;
    }
};

void Serve() {
    /* This function creates a server with specified interceptors, registers the service calls offered by that server, and exposes
	the server over a specified port.
	*/

    // Set the address for the service to host itself on
    std::string address("0.0.0.0:5000"); //ToDo get this from config

    // Add the hosted services 
    ComfortServiceImplementation service;

    // Create a builder instance for the server
    ServerBuilder builder;

    // Add an insecure port to the server builder
    builder.AddListeningPort(address, grpc::InsecureServerCredentials());
    // Register the hosted services to the server (builder)
    builder.RegisterService(&service);

    // Build and start the server, and listen for calls on the specified port
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "Server listening on port: " << address << std::endl;

    // Defer termination for a 'persistent' service
    server->Wait();
}

int main(int argc, char** argv) {
    Serve();

    return 0;
}

File: services/comfortService/comfortService.py
# Standard library imports
import sys
import os
import logging
import math
import time
from concurrent import futures

# Third party imports
import grpc
import yaml

# Local application imports
sys.path.append( os.path.dirname( os.path.dirname( os.path.dirname( os.path.abspath(__file__) ) ) ) )
import protoFiles.python.comfortService.v1.comfort_service_api_v1_pb2 as comfort_service_api_v1_pb2
import protoFiles.python.comfortService.v1.comfort_service_api_v1_pb2_grpc as comfort_service_api_v1_pb2_grpc
import interceptors.python.metricInterceptor as metricInterceptor
import interceptors.python.authenticationInterceptor as authenticationInterceptor
import interceptors.python.rateLimitInterceptor as rateLimitInterceptor

def loadConfigFile(filepath):
	''' This function reads in a YAML configuration file. It takes the relative filepath as an input. It returns a dictionary (?) containing configuration variables.
	'''
	
	with open(os.path.join(sys.path[0], filepath), "r") as f:
		config = yaml.safe_load(f)
		serverConfig = config["server"]

	logging.debug("Succesfully loaded configuration file")
	return serverConfig

def calculateEquivalentVibrationMultiAxis(timeStamps, weightedVibrationX, weightedVibrationY, weightedVibrationZ):
	''' This function takes a time series of weighted vibration data for 3 axes as well as their assosciated timestamps. Using these, it calculates the equivalent vibration magnitude according to SANS 2631-1 (Appendix C, equ C.1).
	'''

	numerator = 0.0 # This variables holds the accumulated numerator use in the calculation of the equivalent vibration magnitude

	denominator = 0.0 # This variables holds the accumulated denominator use in the calculation of the equivalent vibration magnitude

	# Iterate through the request samples
	for i in range(1, len(timeStamps)):
		# Find the time difference between estimates. Assume the acceleration signal is constant over this time period.

		timeGap = timeStamps[i] - timeStamps[i-1]

		# Increment numerator of the equivalent vibration equation.

		numerator += (((weightedVibrationX[i]**2) + (weightedVibrationY[i]**2) + (weightedVibrationZ[i]**2)) * timeGap)

		# Increment denominator of the equivalent vibration equation
		denominator += timeGap

	# Calculate and return the equivalent vibration
	return math.sqrt(numerator/denominator)

def calculateEquivalentVibrationSingleAxis(timeStamps, weightedVibration):
	''' This function takes a time series of weighted vibration data for a single axis as well as their assosciated timestamps. Using these, it calculates the equivalent vibration magnitude according to SANS 2631-1 (Appendix C, equ C.1).
	'''

	numerator = 0.0 # This variables holds the accumulated numerator use in the calculation of the equivalent vibration magnitude

	denominator = 0.0 # This variables holds the accumulated denominator use in the calculation of the equivalent vibration magnitude

	# Iterate through the request samples
	for i in range(1, len(weightedVibration)):
		# Find the time difference between estimates. Assume the acceleration signal is constant over this time period.

		timeGap = timeStamps[i] - timeStamps[i-1]

		# Increment numerator of the equivalent vibration equation.
		numerator += ((weightedVibration[i]**2) * timeGap)

		# Increment denominator of the equivalent vibration equation
		denominator += timeGap

	# Calculate and return the equivalent vibration
	return math.sqrt(numerator/denominator)

def assessComfort(equivalentAcceleration, response):
	''' This function takes an equivalent acceleration and a gRPC response message as inputs. Based on the equivalent acceleration, it sets the rating field of the message according to SANS 2631-1 (Appendix C, section C2.3).
	'''
	
	if(equivalentAcceleration < 0.315):
		response.rating = comfort_service_api_v1_pb2.NOT_UNCOMFORTABLE
	elif(equivalentAcceleration < 0.6):
		response.rating = comfort_service_api_v1_pb2.SLIGHTLY_UNCOMFORTABLE
	elif(equivalentAcceleration < 0.9):
		response.rating = comfort_service_api_v1_pb2.FAIRLY_UNCOMFORTABLE
	elif(equivalentAcceleration < 1.4):
		response.rating = comfort_service_api_v1_pb2.UNCOMFORTABLE
	elif(equivalentAcceleration < 2):
		response.rating = comfort_service_api_v1_pb2.VERY_UNCOMFORTABLE
	else:
		response.rating = comfort_service_api_v1_pb2.EXTREMELY_UNCOMFORTABLE

	return response
		
class ComfortServiceServicer(object):
	"""'Comfort Service' offers one service call that provides information about human comfort onboard, in response to vessel vibrations.
	"""

	def ComfortRating(self, request, context):
		"""The 'Comfort Rating' call provides foresight for tactical decision-making by providing a comfort rating for a proposed route, based on estimated vibrations on board.
		"""

		# print("Starting")
		# startTime = time.time()
		# while ((time.time() - startTime) < 5):
		# 	pass

		responseMessage = comfort_service_api_v1_pb2.ComfortResponse()
		responseMessage.unix_time.extend(request.unix_time)
		
		# Calculate equivalent vibration
		if (len(request.unix_time) > 1):
			equivalentVibration = calculateEquivalentVibrationMultiAxis(request.unix_time, request.human_weighted_vibration_x, request.human_weighted_vibration_y, request.human_weighted_vibration_z)
		else:
			equivalentVibration = math.sqrt((request.human_weighted_vibration_x[0]**2)+(request.human_weighted_vibration_y[0]**2)+(request.human_weighted_vibration_z[0]*2))

		# Assess comfort
		responseMessage = assessComfort(equivalentVibration, responseMessage)

		# print("Sending")
		return responseMessage
	
def loadTLSCredentials(certDirectory):
		# This function loads in the generated TLS credentials from file, creates
	# a server credentials object with the key and certificate, and  returns 
	# that object for use in the server connection
	
	serverKeyFile = f"{certDirectory}/server-key.pem"
	serverCertFile = f"{certDirectory}/server-cert.pem"
	caCertFile = f"{certDirectory}/ca-cert.pem"

	# Load the server's certificate and private key
	private_key = open(serverKeyFile).read()
	certificate_chain = open(serverCertFile).read()

	# Load certificates of the CA who signed the client's certificate
	certificate_pool = open(caCertFile).read()

	credentials = grpc.ssl_server_credentials(
		private_key_certificate_chain_pairs = [(bytes(private_key, 'utf-8'), bytes(certificate_chain, 'utf-8'))],
		root_certificates = certificate_pool,
		require_client_auth = True
	)
	
	return credentials
	
def serve():
	''' This function creates a server with specified interceptors, registers the service calls offered by that server, and exposes
	the server over a specified port.
	'''

	# Create interceptor chain
	activeInterceptors = [
		metricInterceptor.MetricInterceptor("ComfortService"), 
		authenticationInterceptor.AuthenticationInterceptor(
			config["authentication"]["jwt"]["secretKey"], 
			config["authentication"]["jwt"]["tokenDuration"], 
			{config["authentication"]["accessLevel"]["name"]["comfortRating"]: config["authentication"]["accessLevel"]["role"]["comfortRating"]}
		),
		rateLimitInterceptor.RateLimitInterceptor(4)
	] # List containing the interceptors to be chained
	
	# Create a server to serve calls in its own thread
	server = grpc.server(
		futures.ThreadPoolExecutor(max_workers=10),
		interceptors = activeInterceptors
	)
	logging.debug("Successfully created server.")

	# Register an ocean weather service on the server
	comfort_service_api_v1_pb2_grpc.add_ComfortServiceServicer_to_server(ComfortServiceServicer(), server)
	logging.debug("Successfully registered comfort service to server.")

	# Load TLS credentials
	creds = loadTLSCredentials("certification")

	# Create a secure connection on port
	comfortHost = os.getenv(key = "COMFORTHOST", default = "[::]") # Receives the hostname from the environmental variables (for Docker network), or defaults to localhost for local testing
	try:
		# server.add_insecure_port(f'{comfortHost}:{config["port"]["myself"]}')
		server.add_secure_port(f'{comfortHost}:{config["port"]["myself"]}', creds) # server.add_insecure_port(f'{comfortHost}:{config["port"]["myself"]}')
		logging.debug("Succesfully added (insecure) port to server.")
	except Exception as e:
		logging.debug(f"Failed to add (insecure) port to server: \n{e}")

	try:
		# Start server and listen for calls on the specified port
		server.start()
		logging.info(f'Server started on port {config["port"]["myself"]}')

		# Defer termination for a 'persistent' service
		server.wait_for_termination()
	except Exception as e:
		logging.debug(f'Failed to start server on port {config["port"]["myself:"]}: \n{e}')
		
if __name__ == '__main__':
	# ________LOAD CONFIG FILE________
	config = loadConfigFile("configuration.yaml")

	# ________LOGGER SETUP________
	serviceName = __file__.rsplit("/")[-2].rsplit(".")[0]

	logging.basicConfig(filename="services/comfortService/program logs/" + serviceName + ".log", format="%(asctime)s:%(name)s:%(levelname)s:%(module)s:%(funcName)s:%(message)s", level=logging.DEBUG, force = True)

	# ________SERVE REQUESTS________
	serve() # Finish initialisation by serving the request